contributor-book/philosophy_0_80.md

```diff
@@ -0,0 +1,198 @@
+
+Composing commands, both built-in and user-defined, is a core piece of Nushell. The design of Nushell and its standard library must support both easily composing commands as well as allowing the user to easily create compose-able commands.
+### Command philosophy
+
+Specifically, Nushell's philosophy about commands is represented by the following positions (these apply especially to the library of builtin commands distributed with nushell, but are also good guidance for your own commands and plugins):
+
+1. There should be one -- and preferably only one -- obvious way to do it.
+  You will probably recognize this from the "[Zen of Python](https://peps.python.org/pep-0020/)"!  
+
+2. The aim of our library of builtin commands is to provide a collection of simple and composable primitive commands, that covers essential shell programming needs and allows virtually anything to be built via composition and creation of custom user commands and plugins.
+  That's as opposed to maintaining an extensive catalog of more specialized commands in the core nushell codebase: more specialized functionality should exist as community-contributed libraries (e.g. plugins/nushell libraries).
+
+3. If something can be done conveniently by composition of simpler commands then we do not add flags and options to do the same thing.
+
+4. The primary input to a command should generally be supplied as "input", not as a positional argument.
+  This is what enables pipeline composition.
+
+5. A given command, on a given input, should generally always produce the same type of output data structure; the presence/absence/value of arguments and flags shouldn't change the output type.
+## Command signatures and their parts
+
+### Signature
+
+The signature of the command describes the following:
+
+- The name of the command
+- The usage information (documentation) for the command
+- The name and type of:
+    - Named parameters
+    - Positional parameters
+- The type of:
+    - Input
+    - Output
+
+### When to use input
+
+Input is intended for one or both of:
+- Pipeline composition
+- Streams of data
+
+### When to use output
+
+All return values are output.
+
+### When to use positional parameters
+
+Use a positional parameter when:
+- A parameter is required
+
+### When to use rest parameters
+
+Use a rest parameter when:
+- A command takes some number of additional optional arguments of the same type
+
+### When to use named parameters
+
+Use a named parameter (a flag parameter) when:
+- The value you need is optional
+
+### When to use a switch flag
+
+Use a switch flag when:
+- You need to allow the user to change the default action of the command
+
+### Comments
+
+Commands should be commented to describe their function and usage. This documentation should also cover parameters and their use.
+
```

Comment:
```suggestion
## Naming commands

Good naming is incredibly important! Here are some guidelines:

- Find out what names other well-known programming languages use for the thing you are creating (if something similar exists). Don't choose a name that differs from all other languages without a good reason.

- As always in programming, if you are uncomfortable with the name chosen for something, stop! It often means there's something wrong with your design. Maybe it's doing too much? Or maybe the concept it's representing is still a bit confused?

- Don't expect it to be easy -- it may take some time to settle on the right name for a command (and its options, if it needs any). Don't hesitate to come and talk to other Nushell contributors in Discord, or discuss it in GitHub.

- Find out what names Nushell uses for related things -- your name should be harmonious with those. E.g. if `head` already exists, maybe it makes sense for another command called `tail` to exist? (Or maybe the thing called `head` shouldn't have that name! Naming is hard so don't assume current names are perfect or immutable.)

- Remember that when you're choosing a name, you're making a decision that affects the code of every Nushell user in the world. (Of course, this means that fixing bad naming decisions will break the code of Nushell users and force them to update. We still won't always get it right the first time, but this is one reason that we put so much emphasis on it.)

- If the name you feel is the "right name" is already taken, don't just settle for an inferior name -- please start a discussion about the situation instead. It may mean that there is a broader conceptual or naming problem that needs to be addressed.

- Please explain how you chose the name in your PR -- writing this will help with the process of naming it!

The ultimate goal here is to create an ecosystem of commands and terminology that is harmonious, and "feels right". Yes, this is both highly subjective and hard work! But it is an extremely important factor determining how enjoyable and powerful it is to work in the programming environment that we are creating, and ultimately it is this that will determine the success of the project.
```

Reply:
I agree with the sentiment here. Indeed, naming can be difficult, but REnaming is frequently trivial. I'd rather see some awesome wow thing that changes the way we think about things, that is named badly, then have folks struggling for hours trying to come up with the "right" name and never really starting to code. Try to do all the things above but don't stop coding because you don't have the perfect name for something.

Here's a real-world example. zhiburt, who is the author of tabled and a frequent contributor to nushell recently introduced a PR to show off some cool tui things with horizontal and vertical scrolling. The initial name was tabless, kind of clever like table and less combined, I get it. But the point of the PR wasn't the name, it was the total awesomeness that can come with a tui. That thing could've been named bob initially and I would've been equally blown away with it. It truly was a beautiful demo and surprise. I tried to encourage him, and we go back and forth about features and he kept on chugging away. Then someone objected to the name and suggested scroll. zhiburt changed it in a few minutes and scroll is what it is currently named now. Is that the perfect name? I have no idea, but tbh, I'm not really concerned with the exact "right" name. I want to use the tui he introduced whatever its final name ends up being. Having said that, it's sometimes easier to spot the wrong name than to pick the right one. bob would've been the wrong name. lol. And yet again, sometimes other people can just come up with good names, but you can't because you can't see the forest for the trees.

Reply:
I agree with the sentiment here. Indeed, naming can be difficult, but REnaming is frequently trivial. I'd rather see some awesome wow thing that changes the way we think about things, that is named badly, then have folks struggling for hours trying to come up with the "right" name and never really starting to code. Try to do all the things above but don't stop coding because you don't have the perfect name for something.

Here's a real-world example. zhiburt, who is the author of tabled and a frequent contributor to nushell recently introduced a PR to show off some cool tui things with horizontal and vertical scrolling. The initial name was tabless, kind of clever like table and less combined, I get it. But the point of the PR wasn't the name, it was the total awesomeness that can come with a tui. That thing could've been named bob initially and I would've been equally blown away with it. It truly was a beautiful demo and surprise. I tried to encourage him, and we go back and forth about features and he kept on chugging away. Then someone objected to the name and suggested scroll. zhiburt changed it in a few minutes and scroll is what it is currently named now. Is that the perfect name? I have no idea, but tbh, I'm not really concerned with the exact "right" name. I want to use the tui he introduced whatever its final name ends up being. Having said that, it's sometimes easier to spot the wrong name than to pick the right one. bob would've been the wrong name. lol. And yet again, sometimes other people can just come up with good names, but you can't because you can't see the forest for the trees.

Reply:
Thanks @fdncred, I agree we should iterate on these guidelines together. It is true that renaming is trivial when something's not merged. However it's still true that if you are struggling with naming something then that often points to a problem with the design of what you're working on.

But also, what I was getting at (and which I didn't make clear) is that renaming is not at all trivial when something is merged and released. So the purpose of my bullet points was to emphasize that PRs should not be merged until multiple people are happy with the naming. Currently, there are many bad naming and API decisions in the library of Nushell builtins, but those builtins are out there in the wild being used by Nushell users and perhaps even playing a role in working software. So we'll need to make breaking changes to fix this, but we also need to change our own processes so that we get better at releasing builtins that don't need breaking changes to fix design decisions.

Reply:
@dandavison - we'll definitely need to set up naming guidelines so we can get commands names in a predictable/readable shape for the 1.0 release. I think it might take a few revisions to get it how we want it (since I'm sure we even know just yet 😅 )

Can we split the naming piece into its own philosophy doc? That way we can land this one sooner and then add more components to it one after another as they get ready.

Reply:
> Can we split the naming piece into its own philosophy doc?

Absolutely! I just wanted to get some thoughts and draft text down on the subject for others to comment on.